rct229/rulesets/ashrae9012019/data_fns/extra_schema_fns.py

```diff
@@ -0,0 +1,195 @@
+import re
+from functools import partial
+
+from rct229.rulesets.ashrae9012019.data import data
+
+EXTRA_SCHEMA = data["ASHRAE229.9012019.extra.schema"]
+
+exception_list = [
+    "Enumerations2019ASHRAE901",
+    "EnumerationsRESNET",
+    "Enumerations2019T24",
+    "Output2019ASHRAE901",
+]
+
+
+def if_required(required) -> bool:
+    """
+    Convert required data from extra schema to a boolean
+    If the data is tring, then return false else return boolean
+
+    Parameters
+    ----------
+    required: boolean | str
+
+    Returns
+    -------
+    boolean
+
+    """
+    if isinstance(required, str):
+        # unknown
+        return False
+    else:
+        # boolean true or false
+        return required
+
+
+def get_extra_schema_by_data_type(data_type):
+    """
+    Map the referenced data_type to extra schema element
+
+    Parameters
+    ----------
+    data_type: string of data reference
+
+    Returns
+    -------
+    dict | str | None
+
+    """
+    if data_type.startswith("[") or data_type.startswith("{"):
+        # this is a data group
+        data_type = "".join(re.findall(r"[\w\s]+", data_type))
+        if EXTRA_SCHEMA.get(data_type):
+            return EXTRA_SCHEMA[data_type]["Data Elements"]
+    elif data_type.startswith("({"):
+        # this is for a referenced external schema.
+        data_type = "".join(re.findall(r"[\w\s]+", data_type))
+        return data_type
+    return None
+
+
+def compare_context_pair(
+    index_context,
+    compare_context,
+    element_json_path,
+    extra_schema,
+    required_equal,
+    search_key,
+    error_msg_list,
+) -> bool:
+    """
+    Perform equality comparison between two RPD context
+    based on the value in the extra schema specification
+
+    Parameters
+    ----------
+    index_context: dict|list|str|None, the indexed context (RPD)
+    compare_context: dict|list|str|None, the compare context (RPD)
+    element_json_path: str, The json path of the compare context
```

Comment:
An example JSON path might help folks that don't know the convention:

e.g., $.ruleset_model_descriptions[0].buildings[0].building_open_schedule

Reply:
Good point, I can add an example to the comment.